scrapper.py

```diff
@@ -21,43 +36,310 @@ class IncorrectNumberOfArticlesError(Exception):
+
+        self.article.author = first_author_bs.text
+
+        text_date = re.search(r'Поступила в редакцию\s+\d{2}\.\d{2}\.\d{4}', self.article.text)
+        date_re = re.search(r'\d{2}\.\d{2}\.\d{4}', text_date.group(0))
+
+        self.article.date = datetime.strptime(date_re.group(0), '%d.%m.%Y')
+
+
 class Crawler:
     """
     Crawler implementation
     """
+
     def __init__(self, seed_urls, max_articles: int):
-        pass
+        self._seed_urls = seed_urls
+        self.max_articles = max_articles
+        self.urls = []
 
     def _extract_url(self, article_bs):
-        pass
+        """
+        Finds urls from the given article_bs
+        """
+        table_rows_bs = article_bs.find_all('tr', {"class": "unnrow"})
+
+        urls = []
+        overall_urls_count = len(self.urls)
+
+        links_bs = []
+
+        for table_row_bs in table_rows_bs:
+            links_bs.extend(table_row_bs.find_all('a'))
+
+        for link_bs in links_bs:
+            if overall_urls_count + 1 > self.max_articles:
+                break
+
+            # Checks if the link leads to an article
+            match = re.match(r'^\?anum', link_bs['href'])
+
+            if not match:
+                continue
+
+            link_url = ''.join([DOMAIN_URL, link_bs['href']])
+
+            if link_url not in self.urls and link_url not in urls:
+                urls.append(link_url)
+                overall_urls_count += 1
+
+        return urls
 
     def find_articles(self):
         """
         Finds articles
         """
-        pass
+        for seed_url in self._seed_urls:
+            if len(self.urls) + 1 > self.max_articles:
+                break
+
+            time.sleep(random.uniform(0.0, 1.0))
+
+            response = requests.get(seed_url)
+            article_bs = BeautifulSoup(response.text, features="html.parser")
+
+            self.urls.extend(self._extract_url(article_bs))
 
     def get_search_urls(self):
         """
         Returns seed_urls param
         """
-        pass
+        return self._seed_urls
+
+
+class CrawlerRecursive(Crawler):
+    """
+    CrawlerRecursive implementation
+    """
+    def __init__(self, seed_urls, max_articles: int):
+        self.crawled_urls = []
+
+        super().__init__(seed_urls, max_articles)
+
+    def find_articles(self):
+        """
+        Finds articles
+        """
+        start_url = self._seed_urls[0]
+        self._seed_urls = []
+
+        self.urls = load_scrapped_urls()
+        pre_scrapped_url_count = len(self.urls)
+
+        self.crawl(start_url)
+        self.urls = self.urls[pre_scrapped_url_count:]
+
+    def crawl(self, url_to_crawl):
+        if len(self.urls) + 1 > self.max_articles:
+            return
+
+        time.sleep(random.uniform(0.0, 1.0))
+
+        response = requests.get(url_to_crawl)
+        response_bs = BeautifulSoup(response.text, features="html.parser")
+        links_bs = response_bs.find_all('a')
+
+        for link_bs in links_bs:
```

Comment:
regarding my comment about nested code, now I see that this is your current style. Please rework this piece of code as well to make it as shallow as possible

Reply:
Done?